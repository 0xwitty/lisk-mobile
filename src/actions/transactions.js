import actionTypes from '../constants/actions';
import txConstants from '../constants/transactions';
import { getTransactions, send } from '../utilities/api/transactions';
import { loadingStarted, loadingFinished } from './loading';

export const transactionsLoaded = data =>
  (dispatch, getState) => {
    const { activePeer } = getState().peers;
    dispatch(loadingStarted(actionTypes.transactionsLoaded));
    getTransactions(activePeer, data)
      .then((response) => {
        dispatch({
          type: actionTypes.transactionsLoaded,
          data: {
            transactions: response.data,
          },
        });
        dispatch(loadingFinished(actionTypes.transactionsLoaded));
      });
  };

export const transactionAdded = (data, success, error) =>
  (dispatch, getState) => {
    const { activePeer } = getState().peers;
    const account = getState().accounts.active;
    dispatch(loadingStarted(actionTypes.transactionAdded));
    send(activePeer, data)
      .then(({ id }) => {
        dispatch({
          data: {
            id,
            senderPublicKey: account.publicKey,
            senderId: account.address,
            recipientId: data.recipientId,
            amount: data.amount,
            fee: txConstants.send.fee,
            type: txConstants.send.type,
          },
          type: actionTypes.pendingTransactionAdded,
        });
        dispatch(loadingFinished(actionTypes.transactionAdded));
        success({ txId: id });
      }).catch((err) => {
        dispatch(loadingFinished(actionTypes.transactionAdded));
        error(err);
      });
  };
