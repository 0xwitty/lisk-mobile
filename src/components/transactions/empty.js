import React from 'react';
import { View, Image } from 'react-native';
import noActivity from '../../assets/images/noActivity/noActivity2x.png';
import { P } from '../toolBox/typography';
import styles from './styles';

const EmptyState = () =>
  <View style={styles.emptyState}>
    <View style={styles.noActivity}>
      <Image style={styles.image} source={noActivity} />
    </View>
    <P style={styles.noTxTitle}>You do not have any recent activity.</P>
  </View>;

export default EmptyState;
