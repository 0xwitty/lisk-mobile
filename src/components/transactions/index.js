import React, { Fragment } from 'react';
import { View, Animated, Platform, ActivityIndicator } from 'react-native';
import connect from 'redux-connect-decorator';
import ReactNativeHapticFeedback from 'react-native-haptic-feedback';
import {
  settingsUpdated as settingsUpdatedAction,
} from '../../actions/settings';
import List from './list';
import Footer from './footer';
import { H3, Small, A } from '../toolBox/typography';
import { fromRawLsk } from '../../utilities/conversions';
import Icon from '../toolBox/icon';
import colors from '../../constants/styleGuide/colors';
import easing from '../../utilities/easing';
import withTheme from '../withTheme';
import getStyles from './styles';
import { IconButton } from '../toolBox/button';

/**
 * This component is a HOC to decide which state to show:
 * Loading, Empty (No transactions) or the List view.
 *
 * It performs the initial animation after the user logged in.
 *
 */
@connect(state => ({
  incognito: state.settings.incognito,
}), {
  settingsUpdated: settingsUpdatedAction,
})
class Transactions extends React.Component {
  state = {
    initialAnimations: {
      opacity: new Animated.Value(0),
      top: new Animated.Value(20),
    },
  }

  componentDidMount() {
    this.initialFadeIn();
  }

  initialFadeIn = () => {
    const { opacity, top } = this.state.initialAnimations;
    Animated.timing(opacity, {
      toValue: 1,
      duration: 400,
      delay: 100,
    }).start();
    Animated.timing(top, {
      toValue: 0,
      duration: 400,
      delay: 100,
      easing: easing.easeInOutQuart,
    }).start();
  }

  onPress = () => {
    this.props.navigate('Send', { initialize: true });
  }

  toggleIncognito = () => {
    ReactNativeHapticFeedback.trigger('selection');
    this.props.settingsUpdated({
      incognito: !this.props.incognito,
    });
  }

  render() {
    const {
      styles, transactions, navigate,
      account, footer, incognito, theme, placeholderStyle,
      refreshing,
    } = this.props;

    const balance = account ? parseFloat(fromRawLsk(account.balance)) : '';
    const Anim = Animated.View;
    const { opacity, top } = this.state.initialAnimations;

    return (<Anim style={[styles.container, { opacity, top }]}>
      {
        (!transactions ||
          (transactions.confirmed.length === 0 && transactions.pending.length === 0)) ?
          <Fragment></Fragment> :
          <Fragment>
            <View style={[styles.placeholder, placeholderStyle]}>
              {(Platform.OS === 'ios' && refreshing) ? <ActivityIndicator size="large" /> : null}
            </View>
            <View style={styles.innerContainer}>
              <H3 style={[styles.title, styles.theme.title]}>Activity</H3>
              {incognito ?
                <IconButton
                  title=''
                  icon={'disable-incognito'}
                  color={colors.dark.gray2}
                  iconSize={20}
                  onClick={this.toggleIncognito}
                /> : null
              }
            </View>
            {!account.initialized && balance >= 0.2 ?
              <View style={[styles.initContainer, styles.theme.initContainer]}>
                <Icon name='warning' color={colors[theme].red} size={18} />
                <Small style={[styles.initText, styles.theme.initText]}>
                  Your account is not initialized.
                  <A style={[styles.link, styles.theme.link]}
                    onPress={this.onPress}> Initialize it now</A>
                </Small>
              </View> : null
            }
            <List
              incognito={incognito}
              navigate={navigate}
              account={account ? account.address : ''}
              pending={transactions.pending}
              transactions={transactions.confirmed} />
            {
              footer ? <Footer /> : null
            }
          </Fragment>
      }
    </Anim>);
  }
}
export default withTheme(Transactions, getStyles());
