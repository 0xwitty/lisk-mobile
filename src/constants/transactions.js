export default {
  send: {
    type: 0,
    fee: 1e7,
  },
  setSecondPassphrase: {
    type: 1,
    fee: 5e8,
  },
  registerDelegate: {
    type: 2,
    fee: 25e8,
  },
  vote: {
    type: 3,
    fee: 1e8,
  },
};
