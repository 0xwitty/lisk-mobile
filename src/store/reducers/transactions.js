import actionTypes from '../../constants/actions';

/**
 * The reducer of transactions.
 * this reducer implements the required logic to add, update and reset
 * the list of transactions for the account we've logged in using.
 *
 * @param {Object} state
 * @param {Array} state.pending
 * @param {Array} state.confirmed
 * @param {Number} state.count
 * @param {Object} action
 * @param {String} action.type
 * @param {Object} action.data
 */
const transactions = (state = { pending: [], confirmed: [], count: null }, action) => {
  switch (action.type) {
    case actionTypes.transactionsLoaded:
      return Object.assign({}, state, {
        confirmed: [
          ...state.confirmed,
          ...action.data.transactions,
        ],
        count: action.data.count,
      });
    case actionTypes.transactionAdded:
      return Object.assign({}, state, {
        pending: [action.data, ...state.pending],
      });
    default:
      return state;
  }
};

export default transactions;
