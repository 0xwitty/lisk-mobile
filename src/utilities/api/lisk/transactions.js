import * as Lisk from '@liskhq/lisk-client';
import { apiClient } from './apiClient';
import {
  isTransfer,
  moduleAssetNameIdMap,
  transferAssetSchema
} from '../../../constants/transactions';
import config from '../../../../lsk.config';

const getAmount = (tx) => {
  if (moduleAssetNameIdMap.transfer === tx.moduleAssetId) {
    return tx.asset.amount;
  }
  if (moduleAssetNameIdMap.unlockToken === tx.moduleAssetId) {
    // eslint-disable-next-line no-undef
    return tx.asset?.unlockObjects?.reduce?.((a, b) => a + BigInt(b.amount), BigInt(0)).toString();
  }
  return 0;
};

/**
 * Normalizes transaction data retrieved from Lisk Core API
 * https://lisk.com/documentation/lisk-core/reference/api.html#/Transactions/get_transactions__id_
 */
const normalizeTransactionsResponse = (list, block) =>
  list.map((tx) => ({
    id: tx.id,
    senderAddress: tx.sender.address,
    recipientAddress: isTransfer({ moduleAssetId: tx.moduleAssetId })
      ? tx.asset.recipient.address
      : tx.sender.address,
    amount: getAmount(tx),
    fee: tx.fee,
    timestamp: tx.block?.timestamp,
    confirmations: block ? block.height - tx.block?.height + 1 : 0,
    nonce: tx.nonce,
    type: tx.moduleAssetName,
    moduleAssetId: tx.moduleAssetId,
    moduleAssetName: tx.moduleAssetName,
    data: tx.moduleAssetName === 'token:transfer' ? tx.asset.data : '',
    votes: tx.moduleAssetName === 'dpos:voteDelegate' ? tx.asset.votes : [],
    delegate: tx.moduleAssetName === 'dpos:registerDelegate' ? tx.asset.username : '',
    blockId: tx.block?.id,
    blockHeight: tx.block?.height,
  }));

export const get = async ({
  id, address, limit, offset
}) => {
  const block = await apiClient.getLatestBlock();
  if (id !== undefined) {
    const txs = await apiClient.getTransaction(id);
    return { data: normalizeTransactionsResponse(txs, block), meta: {} };
  }
  const txs = await apiClient.getTransactions(address, limit, offset);
  return {
    data: normalizeTransactionsResponse(txs, block),
    meta: {
      limit,
      offset,
      count: txs.length
    }
  };
};

export const create = async ({
  nonce, passphrase, recipientAddress, amount, fee, reference
}) => {
  const {
    publicKey: senderPublicKey
  } = Lisk.cryptography.getAddressAndPublicKeyFromPassphrase(passphrase);
  const recipient = Lisk.cryptography.getAddressFromLisk32Address(recipientAddress);
  const tx = {
    moduleID: 2,
    assetID: 0,
    senderPublicKey,
    // eslint-disable-next-line no-undef
    fee: BigInt(fee),
    // eslint-disable-next-line no-undef
    nonce: BigInt(nonce),
    asset: {
      // eslint-disable-next-line no-undef
      amount: BigInt(amount),
      data: reference || '',
      recipientAddress: recipient
    },
    signatures: []
  };

  const networkIdentifier = Buffer.from(
    config.isTestnet ? config.testnetNetworkID : config.networkID,
    'hex'
  );
  return Lisk.transactions.signTransaction(transferAssetSchema, tx, networkIdentifier, passphrase);
};

export const broadcast = async (transaction) => {
  const txBytes = Lisk.transactions.getBytes(transferAssetSchema, transaction).toString('hex');
  return apiClient.sendTransaction({ transaction: txBytes });
};
