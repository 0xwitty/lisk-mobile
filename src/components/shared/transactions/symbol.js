import React from 'react';
import { View, Image } from 'react-native';
import Avatar from '../avatar';
import Icon from '../toolBox/icon';
import { getTxConstant, moduleAssetNameIdMap } from '../../../constants/transactions';
import withTheme from '../withTheme';
import getStyles from './styles';
import { colors, themes } from '../../../constants/styleGuide';

const Symbol = ({
  token,
  theme,
  direction,
  sender,
  recipient,
  address,
  styles,
  moduleAssetId,
}) => {
  if (token !== 'LSK') {
    return (
      <View
        style={[styles.transactionIcon, styles.theme[`${direction}Symbol`]]}
      >
        <Icon
          name={direction}
          size={14}
          color={
            direction === 'outgoing'
              ? colors[theme].outgoingArrow
              : colors[theme].ufoGreen
          }
        />
      </View>
    );
  }
  if (moduleAssetId === moduleAssetNameIdMap.transfer && sender !== recipient) {
    return <Avatar address={address} size={40} style={styles.theme.avatar} />;
  }
  return (
    <Image
      source={getTxConstant({ moduleAssetId })?.image?.(themes[theme])}
      style={styles.image}
    />
  );
};

export default withTheme(Symbol, getStyles());
