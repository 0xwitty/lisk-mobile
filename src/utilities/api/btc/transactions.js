import bitcoin from 'bitcoinjs-lib';
import config from '../../../../btc.config';
import { extractAddress, getDerivedPathFromPassphrase } from './account';
import { merge } from '../../helpers';
import { validateAddress } from '../../validators';
import { tokenMap } from '../../../constants/tokens';

/**
 * Normalizes transaction data retrieved from Blockchain.info API
 * @param {Object} data
 * @param {String} data.address Base address to use for formatting transactions
 * @param {Array} data.list Transaction list retrieved from API
 * @param {Number} data.blockHeight Latest block height for calculating confirmation count
 */
const normalizeTransactionsResponse = ({
  address,
  list,
  blockHeight,
}) => list.map((tx) => {
  const data = {
    id: tx.hash,
    timestamp: Number(tx.time) * 1000,
    confirmations: blockHeight > 0 ? (blockHeight - tx.block_height) + 1 : tx.block_height,
    type: 0,
    data: '',
  };

  const totalInput = tx.inputs.reduce((total, t) => total + t.prev_out.value, 0);
  const totalOutput = tx.out.reduce((total, t) => total + t.value, 0);
  data.fee = totalInput - totalOutput;

  const ownedInput = tx.inputs.find(i => i.prev_out.addr === address);

  if (ownedInput) {
    data.senderAddress = address;
    const extractedAddress = tx.out[0].addr;
    data.recipientAddress = validateAddress(tokenMap.BTC.key, extractedAddress) === 0 ? extractedAddress : 'Unparsed Address';
    data.amount = tx.out[0].value;
  } else {
    const output = tx.out.find(out => out.addr === address);
    const extractedAddress = tx.inputs[0].prev_out.addr;
    data.senderAddress = validateAddress(tokenMap.BTC.key, extractedAddress) === 0 ? extractedAddress : 'Unparsed Address';
    data.recipientAddress = address;
    data.amount = output.value;
  }

  return data;
});

/**
 * Retrieves latest block from the Blockchain.info API and returns the height.
 * @returns {Promise<Number>}
 */
export const getLatestBlockHeight = () => new Promise(async (resolve) => {
  try {
    const response = await fetch(`${config.url}/latestblock`);
    const json = await response.json();

    if (response.ok) {
      resolve(json.height);
    } else {
      resolve(0);
    }
  } catch (error) {
    resolve(0);
  }
});

export const get = ({
  id,
  address,
  limit = 50,
  offset = 0,
}) => new Promise(async (resolve, reject) => {
  try {
    let response;

    if (id) {
      response = await fetch(`${config.url}/rawtx/${id}`, config.requestOptions);
    } else {
      response = await fetch(`${config.url}/rawaddr/${address}?limit=${limit}&offset=${offset}`, config.requestOptions);
    }

    const json = await response.json();

    if (response.ok) {
      const blockHeight = await exports.getLatestBlockHeight();

      resolve({
        data: normalizeTransactionsResponse({
          address,
          list: id ? [json] : json.txs,
          blockHeight,
        }),
        meta: {
          count: id ? 1 : json.n_tx,
        },
      });
    } else {
      reject(json);
    }
  } catch (error) {
    reject(error);
  }
});

/**
 * Normalizes transaction data retrieved from Blockchain.info API
 * @param {Object} data
 * @param {Number} data.inputCount
 * @param {Number} data.outputCount
 * @param {Number} data.dynamicFeePerByte - in satoshis/byte.
 */
export const calculateTransactionFee = ({
  inputCount,
  outputCount,
  dynamicFeePerByte,
}) => ((inputCount * 180) + (outputCount * 34) + 10 + inputCount) * dynamicFeePerByte;

/**
 * Retrieves unspent tx outputs of a BTC address from Blockchain.info API
 * @param {String} address
 * @returns {Promise<Array>}
 */
export const getUnspentTransactionOutputs = address => new Promise(async (resolve, reject) => {
  try {
    const response = await fetch(`${config.url}/unspent?active=${address}`);
    const json = await response.json();

    if (response.ok) {
      resolve(json.unspent_outputs);
    } else {
      reject(json);
    }
  } catch (error) {
    reject(error);
  }
});

export const create = ({
  passphrase,
  recipientAddress,
  amount,
  dynamicFeePerByte,
}) => new Promise(async (resolve, reject) => {
  try {
    amount = Number(amount);
    dynamicFeePerByte = Number(dynamicFeePerByte);

    const senderAddress = extractAddress(passphrase);
    const unspentTxOuts = await exports.getUnspentTransactionOutputs(senderAddress);

    // Estimate total cost (currently estimates max cost by assuming the worst case)
    const estimatedMinerFee = calculateTransactionFee({
      inputCount: unspentTxOuts.length,
      outputCount: 2,
      dynamicFeePerByte,
    });

    const estimatedTotal = amount + estimatedMinerFee;

    // Check if balance is sufficient
    const unspentTxOutsTotal = unspentTxOuts.reduce((total, tx) => {
      total += tx.value;
      return total;
    }, 0);

    if (unspentTxOutsTotal < estimatedTotal) {
      reject(new Error('Insufficient (estimated) balance'));
    }

    // Find unspent txOuts to spend for this tx
    let txOutIndex = 0;
    let sumOfConsumedOutputs = 0;
    const txOutsToConsume = [];

    while (sumOfConsumedOutputs <= estimatedTotal) {
      const tx = unspentTxOuts[txOutIndex];
      txOutsToConsume.push(tx);
      txOutIndex += 1;
      sumOfConsumedOutputs += tx.value;
    }

    const txb = new bitcoin.TransactionBuilder(config.network);

    // Add inputs from unspent txOuts
    // eslint-disable-next-line
    for (const tx of txOutsToConsume) {
      txb.addInput(tx.tx_hash_big_endian, tx.tx_output_n);
    }

    // Output to Recipient
    txb.addOutput(recipientAddress, amount);

    // Calculate final fee
    const calculatedMinerFee = calculateTransactionFee({
      inputCount: txOutsToConsume.length,
      outputCount: 2,
      dynamicFeePerByte,
    });

    // Calculate total
    const calculatedTotal = amount + calculatedMinerFee;

    // Output to Change Address
    const change = sumOfConsumedOutputs - calculatedTotal;
    txb.addOutput(senderAddress, change);

    // Sign inputs
    const derivedPath = getDerivedPathFromPassphrase(passphrase);
    const keyPair = bitcoin.ECPair.fromWIF(derivedPath.toWIF(), config.network);
    for (let i = 0; i < txOutsToConsume.length; i++) {
      txb.sign(i, keyPair);
    }

    resolve(txb.build().toHex());
  } catch (error) {
    reject(error);
  }
});

export const broadcast = transaction => new Promise(async (resolve, reject) => {
  try {
    const body = new FormData();
    body.append('tx', transaction);

    const response = await fetch(`${config.url}/pushtx`, merge(config.requestOptions, {
      method: 'POST',
      body,
    }));

    if (response.ok) {
      resolve(response.body);
    } else {
      reject(response.body);
    }
  } catch (error) {
    reject(error);
  }
});
