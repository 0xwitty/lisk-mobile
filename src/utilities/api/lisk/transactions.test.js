import { getTransactions, send } from './transactions';
import LiskAPIClient from './apiClient';

LiskAPIClient.transactions.get = jest.fn();
LiskAPIClient.transactions.broadcast = jest.fn();

const transactions = [{
  id: '13746538517771550559',
  height: 6174907,
  blockId: '10687269197653491500',
  type: 0,
  timestamp: 72473846,
  senderPublicKey: 'b550ede5ee0373d2930525016a45aadb50ab6f776068eef3874fa813a26c78d8',
  senderId: '7056261880661230236L',
  recipientId: '4913610097661759129L',
  recipientPublicKey: 'f6399b0251eb0e175e55fac0d6bd53f091c732931dbd816b4bfc42b158d5f04f',
  amount: '10000000000',
  fee: '10000000',
  signature: '30ab885c21e2182d42d5f88ec2ed87dbbaab2a47ac1ccb22d6dc1013dd6c6abc9f9d1d867b554ec3486e70760d544e5948e3040f6ae33bb125ffd3dceedd1901',
  signatures: [],
  confirmations: 15227,
  asset: {
    data: 'manu',
  },
}];

describe('api/transactions', () => {
  it('calls transactions.get method correctly', () => {
    getTransactions({ offset: 0 });
    expect(LiskAPIClient.transactions.get).toBeCalledWith({ offset: 0, sort: 'timestamp:desc' });
  });

  it('calls transactions.broadcast method correctly', async () => {
    LiskAPIClient.transactions.broadcast.mockResolvedValueOnce(transactions[0]);
    await send(transactions[0]);
    expect(LiskAPIClient.transactions.broadcast).toBeCalled();
  });

  it('handles rejections from transactions.broadcast method', async () => {
    const errorMessage = { message: 'Error!' };
    LiskAPIClient.transactions.broadcast.mockRejectedValueOnce(errorMessage);

    try {
      await send(transactions[0]);
    } catch (error) {
      expect(error).toEqual(errorMessage);
    }
  });
});
