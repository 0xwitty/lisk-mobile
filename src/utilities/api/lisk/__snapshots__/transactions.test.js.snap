// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/transactions Should call transactions.broadcast method correctly 1`] = `
Object {
  "amount": "1000000",
  "asset": Object {},
  "block": Object {
    "timestamp": 72473846,
  },
  "blockId": "9937261865542489722",
  "confirmations": 99170,
  "fee": "10000000",
  "height": 7450256,
  "id": "10703320139793065364",
  "recipient": Object {
    "address": "lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw",
  },
  "recipientId": "lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw",
  "recipientPublicKey": "04ffd6fb4508a5ca85355cc3b30dce326fe35781d32deb858879128cf8c141cc",
  "sender": Object {
    "address": "lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw",
  },
  "senderId": "lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw",
  "senderPublicKey": "84795ae68a1908d2a89daafce3f9980811ddf8f0c087c913ebe57759c977fbbc",
  "signature": "08d2526bcd49299bb6894fb62a39b5d084c132cb593be674096aad8b273190922f24677def02817b424c7db1130f14799a0ffc4f95744980f2120c0f1376e200",
  "signatures": Array [],
  "timestamp": 85447190,
  "type": 0,
}
`;

exports[`api/transactions Should call transactions.create method correctly 1`] = `
Object {
  "asset": Object {
    "amount": 1000000n,
    "data": "",
    "recipientAddress": Object {
      "data": Array [
        177,
        180,
        208,
        122,
        157,
        2,
        172,
        70,
        211,
        86,
        194,
        96,
        254,
        186,
        1,
        204,
        183,
        176,
        135,
        71,
      ],
      "type": "Buffer",
    },
  },
  "assetID": 0,
  "fee": 100n,
  "id": Object {
    "data": Array [
      229,
      210,
      78,
      55,
      121,
      86,
      167,
      182,
      2,
      148,
      243,
      138,
      11,
      35,
      236,
      39,
      188,
      68,
      156,
      207,
      241,
      52,
      34,
      47,
      20,
      98,
      171,
      113,
      135,
      248,
      161,
      58,
    ],
    "type": "Buffer",
  },
  "moduleID": 2,
  "nonce": 5n,
  "senderPublicKey": Object {
    "data": Array [
      103,
      211,
      181,
      234,
      240,
      192,
      191,
      107,
      90,
      96,
      45,
      53,
      157,
      174,
      204,
      134,
      167,
      167,
      64,
      83,
      73,
      14,
      195,
      122,
      224,
      142,
      113,
      54,
      5,
      135,
      200,
      112,
    ],
    "type": "Buffer",
  },
  "signatures": Array [
    Object {
      "data": Array [
        190,
        84,
        226,
        6,
        30,
        122,
        189,
        156,
        181,
        50,
        122,
        202,
        130,
        159,
        126,
        129,
        237,
        206,
        175,
        255,
        164,
        178,
        54,
        234,
        211,
        250,
        76,
        111,
        86,
        3,
        217,
        60,
        62,
        75,
        248,
        248,
        127,
        113,
        52,
        43,
        155,
        250,
        171,
        78,
        73,
        77,
        142,
        44,
        225,
        251,
        186,
        18,
        43,
        78,
        215,
        97,
        246,
        228,
        17,
        167,
        164,
        158,
        213,
        11,
      ],
      "type": "Buffer",
    },
  ],
}
`;

exports[`api/transactions Should handle rejections from transactions.broadcast method 1`] = `[TypeError: Cannot mix BigInt and other types, use explicit conversions]`;
